mmcv/ops/csrc/pytorch/cuda/nattenav.cu

```diff
@@ -0,0 +1,1388 @@
+        LAUNCH_DNA_KNS(kernel_size, nattenv_cuda_backward_kernel_fp32,
+                       grid_value, block, 0, stream, d_out_a, d_value_a, attn_a,
+                       height, width, heads, dim, n_value);
+      }));
+  return {d_attn, d_value};
+}
+
+std::vector<torch::Tensor> nattenav_cuda_backward_fp16(
+    const torch::Tensor &d_out, const torch::Tensor &attn,
+    const torch::Tensor &value) {
+  int64_t batch_size = value.size(0);
+  int64_t heads = value.size(1);
+  int64_t height = value.size(2);
+  int64_t width = value.size(3);
+  int64_t dimhalf = value.size(4) / 2;
+  TORCH_CHECK(dimhalf * 2 == value.size(4),
+              "Dims per head must be an even number in FP16.");
+  int kernel_size_sq = attn.size(4);
+  int kernel_size = sqrt(kernel_size_sq);
+  int zsize = batch_size * heads;
+  int xsize = height * width;
+  CHECK_FEATMAP(height, width, kernel_size);
+  CHECK_KERNELSIZE("nattenav_cuda_backward_fp16", kernel_size);
+  int KERNELTHREADS = min(CUDA_NUM_THREADS, kernel_size_sq);
+  int PIXELTHREADS = min(int(CUDA_NUM_THREADS / KERNELTHREADS), xsize);
+  int BATCHTHREADS = max(1, CUDA_NUM_THREADS / (PIXELTHREADS * KERNELTHREADS));
+
+  auto d_attn = torch::zeros_like(attn);
+  auto d_value = torch::zeros_like(value);
+
+  const dim3 attn_blocks((xsize + PIXELTHREADS - 1) / PIXELTHREADS,
+                         (kernel_size_sq + KERNELTHREADS - 1) / KERNELTHREADS,
+                         (zsize + BATCHTHREADS - 1) / BATCHTHREADS);
+  const dim3 attn_threads(PIXELTHREADS, KERNELTHREADS, BATCHTHREADS);
+  int32_t nhalf_value = d_value.numel() / 2;
+  int blocks_value = GET_BLOCKS(nhalf_value);
+  dim3 grid_value(blocks_value);
+  dim3 block(CUDA_NUM_THREADS);
+  const auto stream = c10::cuda::getCurrentCUDAStream();
+  AT_DISPATCH_HALF_TYPES(
+      at::kHalf, d_attn.scalar_type(), "nattenav_backward_cuda_fp16", ([&] {
+        auto d_attn_a =
+            d_attn.packed_accessor32<scalar_t, 5, torch::DefaultPtrTraits>();
+        auto d_value_a =
+            d_value.packed_accessor32<scalar_t, 5, torch::DefaultPtrTraits>();
+        const auto d_out_a =
+            d_out.packed_accessor32<scalar_t, 5, torch::DefaultPtrTraits>();
+        const auto value_a =
+            value.packed_accessor32<scalar_t, 5, torch::DefaultPtrTraits>();
+        const auto attn_a =
+            attn.packed_accessor32<scalar_t, 5, torch::DefaultPtrTraits>();
+        LAUNCH_DNA_KNS(kernel_size, nattena_cuda_backward_kernel_fp16,
+                       attn_blocks, attn_threads, 0, stream, d_out_a, d_attn_a,
+                       value_a, height, width, batch_size, heads, dimhalf);
+        LAUNCH_DNA_KNS(kernel_size, nattenv_cuda_backward_kernel_fp16,
+                       grid_value, block, 0, stream, d_out_a, d_value_a, attn_a,
+                       height, width, heads, dimhalf, nhalf_value);
+      }));
+  return {d_attn, d_value};
+}
+
+// C++ interface
+#define CHECK_CUDA(x) \
```

Comment:
We have defined these macro in common/pytorch_cpp_helper.hpp

Reply:
This definition has been removed.